support for no_std

Any chance you can support `#![no_std]`?

I'd like to use your crate within SGX, using https://github.com/baidu/rust-sgx-sdk

If I build your no-std branch with std support, I get:
```
error[E0433]: failed to resolve: maybe a missing `extern crate std_support;`?
   --> /home/abrenzikofer/.cargo/git/checkouts/twox-hash-d1ab4284274caa19/68108c8/src/thirty_two.rs:219:11
    |
219 | pub use ::std_support::thirty_two::RandomXxHashBuilder;
    |           ^^^^^^^^^^^ maybe a missing `extern crate std_support;`?
error[E0433]: failed to resolve: maybe a missing `extern crate std_support;`?
   --> /home/abrenzikofer/.cargo/git/checkouts/twox-hash-d1ab4284274caa19/68108c8/src/lib.rs:289:9
    |
289 | pub use std_support::sixty_four::RandomXxHashBuilder;
    |         ^^^^^^^^^^^ maybe a missing `extern crate std_support;`?
error[E0432]: unresolved import
  --> /home/abrenzikofer/.cargo/git/checkouts/twox-hash-d1ab4284274caa19/68108c8/src/lib.rs:46:9
   |
46 | pub use thirty_two::RandomXxHashBuilder as RandomXxHashBuilder32;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

your branch `no-std` currently fails to build with `std`:

```
error[E0433]: failed to resolve: maybe a missing `extern crate std_support;`?
   --> src/thirty_two.rs:219:11
    |
219 | pub use ::std_support::thirty_two::RandomXxHashBuilder;
    |           ^^^^^^^^^^^ maybe a missing `extern crate std_support;`?

error[E0433]: failed to resolve: maybe a missing `extern crate std_support;`?
   --> src/lib.rs:289:9
    |
289 | pub use std_support::sixty_four::RandomXxHashBuilder;
    |         ^^^^^^^^^^^ maybe a missing `extern crate std_support;`?

error[E0432]: unresolved import
  --> src/lib.rs:46:9
   |
46 | pub use thirty_two::RandomXxHashBuilder as RandomXxHashBuilder32;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -15,3 +15,4 @@ script: |
   fi
   cargo build --verbose
   cargo test --verbose
+  ci/check-no-std.sh

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -14,12 +14,13 @@ documentation = "https://docs.rs/twox-hash/"
 license = "MIT"
 
 [dependencies]
-rand = ">= 0.3.10, < 0.7"
+rand = { version = ">= 0.3.10, < 0.7", optional = true }
 serde = { version = "1.0", features = ["derive"], optional = true}
 
 [dev-dependencies]
 serde_json = "1.0"
 
 [features]
+default = ["std"]
 serialize = ["serde"]
-
+std = ["rand"]

--- a/ci/check-no-std.sh
+++ b/ci/check-no-std.sh
@@ -1,0 +1,29 @@
+#!/usr/bin/env bash
+
+set -eu
+
+# Investigate `cargo nono` when it's more mature
+# https://github.com/hobofan/cargo-nono/issues/26
+
+echo "Checking for stdlib symbols"
+
+cargo build
+
+symbols=$(nm target/debug/libtwox_hash.rlib 2> /dev/null || true)
+
+if [[ $symbols != *"std"* ]]; then
+    echo 'Did not find symbols that look like the standard library'
+    exit 1
+fi
+
+echo "Checking for no stdlib symbols"
+
+cargo build --no-default-features
+
+symbols=$(nm target/debug/libtwox_hash.rlib 2> /dev/null || true)
+
+if [[ $symbols == *"std"* ]]; then
+    echo 'Found symbols that look like the standard library:'
+    grep std <<< "${symbols}"
+    exit 1
+fi

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -25,6 +25,12 @@
 //! assert_eq!(hash.get(&42), Some(&"the answer"));
 //! ```
 
+#![no_std]
+
+#[cfg(test)]
+extern crate std;
+
+#[cfg(feature = "std")]
 extern crate rand;
 
 #[cfg(feature="serialize")]
@@ -36,10 +42,10 @@ mod number_streams;
 mod thirty_two;
 
 pub use thirty_two::XxHash as XxHash32;
+#[cfg(feature = "std")]
 pub use thirty_two::RandomXxHashBuilder as RandomXxHashBuilder32;
 
-use std::hash::{Hasher, BuildHasher};
-use rand::Rng;
+use core::hash::Hasher;
 use number_streams::NumberStreams;
 
 const CHUNK_SIZE: usize = 32;
@@ -149,8 +155,8 @@ impl XxCore {
     }
 }
 
-impl std::fmt::Debug for XxCore {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+impl core::fmt::Debug for XxCore {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
         write!(
             f, "XxCore {{ {:016x} {:016x} {:016x} {:016x} }}",
             self.v1, self.v2, self.v3, self.v4
@@ -165,7 +171,7 @@ impl XxHash {
             total_len: 0,
             seed: seed,
             core: XxCore::with_seed(seed),
-            buffer: unsafe { ::std::mem::uninitialized() },
+            buffer: unsafe { ::core::mem::uninitialized() },
             buffer_usage: 0,
         }
     }
@@ -188,7 +194,7 @@ impl Hasher for XxHash {
         if self.buffer_usage + bytes.len() < self.buffer.len() {
             unsafe {
                 let tail = self.buffer.as_mut_ptr().offset(self.buffer_usage as isize);
-                std::ptr::copy_nonoverlapping(bytes.as_ptr(), tail, bytes.len());
+                core::ptr::copy_nonoverlapping(bytes.as_ptr(), tail, bytes.len());
             }
             self.buffer_usage += bytes.len();
             return;
@@ -202,7 +208,7 @@ impl Hasher for XxHash {
 
             unsafe {
                 let tail = self.buffer.as_mut_ptr().offset(self.buffer_usage as isize);
-                std::ptr::copy_nonoverlapping(to_use.as_ptr(), tail, bytes_to_use);
+                core::ptr::copy_nonoverlapping(to_use.as_ptr(), tail, bytes_to_use);
             }
 
             let (iter, _) = self.buffer.u64_stream();
@@ -220,7 +226,7 @@ impl Hasher for XxHash {
         // Save any leftover data for the next call
         if bytes.len() > 0 {
             unsafe {
-                std::ptr::copy_nonoverlapping(bytes.as_ptr(), self.buffer.as_mut_ptr(), bytes.len());
+                core::ptr::copy_nonoverlapping(bytes.as_ptr(), self.buffer.as_mut_ptr(), bytes.len());
             }
             self.buffer_usage = bytes.len();
         }
@@ -279,28 +285,16 @@ impl Hasher for XxHash {
     }
 }
 
-#[derive(Clone)]
-/// Constructs a randomized seed and reuses it for multiple hasher instances.
-pub struct RandomXxHashBuilder(u64);
+#[cfg(feature = "std")]
+mod std_support;
 
-impl RandomXxHashBuilder {
-    fn new() -> RandomXxHashBuilder {
-        RandomXxHashBuilder(rand::thread_rng().gen())
-    }
-}
-
-impl Default for RandomXxHashBuilder {
-    fn default() -> RandomXxHashBuilder { RandomXxHashBuilder::new() }
-}
-
-impl BuildHasher for RandomXxHashBuilder {
-    type Hasher = XxHash;
-
-    fn build_hasher(&self) -> XxHash { XxHash::with_seed(self.0) }
-}
+#[cfg(feature = "std")]
+pub use std_support::sixty_four::RandomXxHashBuilder;
 
 #[cfg(test)]
 mod test {
+    use std::prelude::v1::*;
+
     use std::hash::{Hasher, BuildHasherDefault};
     use std::collections::HashMap;
     use super::{XxHash, RandomXxHashBuilder};

--- a/src/number_streams.rs
+++ b/src/number_streams.rs
@@ -1,5 +1,5 @@
-use std::marker::PhantomData;
-use std::ptr;
+use core::marker::PhantomData;
+use core::ptr;
 
 // We should use {u32,u64}::BYTES when they are stable;
 const U32_BYTES: usize = 4;
@@ -107,6 +107,7 @@ impl<'a> NumberStreams for [u8] {
 
 #[cfg(test)]
 mod test {
+    use std::prelude::v1::*;
     use ::std::slice;
     use super::{U32_BYTES,U64_BYTES};
     use super::NumberStreams;

--- a/src/std_support.rs
+++ b/src/std_support.rs
@@ -1,0 +1,51 @@
+pub mod sixty_four {
+    use rand::{self, Rng};
+    use core::hash::BuildHasher;
+    use ::XxHash;
+
+    #[derive(Clone)]
+    /// Constructs a randomized seed and reuses it for multiple hasher instances.
+    pub struct RandomXxHashBuilder(u64);
+
+    impl RandomXxHashBuilder {
+        fn new() -> RandomXxHashBuilder {
+            RandomXxHashBuilder(rand::thread_rng().gen())
+        }
+    }
+
+    impl Default for RandomXxHashBuilder {
+        fn default() -> RandomXxHashBuilder { RandomXxHashBuilder::new() }
+    }
+
+    impl BuildHasher for RandomXxHashBuilder {
+        type Hasher = XxHash;
+
+        fn build_hasher(&self) -> XxHash { XxHash::with_seed(self.0) }
+    }
+}
+
+pub mod thirty_two {
+    use rand::{self, Rng};
+    use core::hash::BuildHasher;
+    use ::thirty_two::XxHash;
+
+    #[derive(Clone)]
+    /// Constructs a randomized seed and reuses it for multiple hasher instances. See the usage warning on `XxHash32`.
+    pub struct RandomXxHashBuilder(u32);
+
+    impl RandomXxHashBuilder {
+        fn new() -> RandomXxHashBuilder {
+            RandomXxHashBuilder(rand::thread_rng().gen())
+        }
+    }
+
+    impl Default for RandomXxHashBuilder {
+        fn default() -> RandomXxHashBuilder { RandomXxHashBuilder::new() }
+    }
+
+    impl BuildHasher for RandomXxHashBuilder {
+        type Hasher = XxHash;
+
+        fn build_hasher(&self) -> XxHash { XxHash::with_seed(self.0) }
+    }
+}

--- a/src/thirty_two.rs
+++ b/src/thirty_two.rs
@@ -1,8 +1,6 @@
-use std;
-use rand;
+use core;
 
-use std::hash::{Hasher, BuildHasher};
-use rand::Rng;
+use core::hash::Hasher;
 use number_streams::NumberStreams;
 
 const CHUNK_SIZE: usize = 16;
@@ -100,8 +98,8 @@ impl XxCore {
     }
 }
 
-impl std::fmt::Debug for XxCore {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+impl core::fmt::Debug for XxCore {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
         write!(
             f, "XxCore {{ {:016x} {:016x} {:016x} {:016x} }}",
             self.v1, self.v2, self.v3, self.v4
@@ -116,7 +114,7 @@ impl XxHash {
             total_len: 0,
             seed: seed,
             core: XxCore::with_seed(seed),
-            buffer: unsafe { ::std::mem::uninitialized() },
+            buffer: unsafe { ::core::mem::uninitialized() },
             buffer_usage: 0,
         }
     }
@@ -139,7 +137,7 @@ impl Hasher for XxHash {
         if self.buffer_usage + bytes.len() < self.buffer.len() {
             unsafe {
                 let tail = self.buffer.as_mut_ptr().offset(self.buffer_usage as isize);
-                std::ptr::copy_nonoverlapping(bytes.as_ptr(), tail, bytes.len());
+                core::ptr::copy_nonoverlapping(bytes.as_ptr(), tail, bytes.len());
             }
             self.buffer_usage += bytes.len();
             return;
@@ -153,7 +151,7 @@ impl Hasher for XxHash {
 
             unsafe {
                 let tail = self.buffer.as_mut_ptr().offset(self.buffer_usage as isize);
-                std::ptr::copy_nonoverlapping(to_use.as_ptr(), tail, bytes_to_use);
+                core::ptr::copy_nonoverlapping(to_use.as_ptr(), tail, bytes_to_use);
             }
 
             let (iter, _) = self.buffer.u32_stream();
@@ -171,7 +169,7 @@ impl Hasher for XxHash {
         // Save any leftover data for the next call
         if bytes.len() > 0 {
             unsafe {
-                std::ptr::copy_nonoverlapping(bytes.as_ptr(), self.buffer.as_mut_ptr(), bytes.len());
+                core::ptr::copy_nonoverlapping(bytes.as_ptr(), self.buffer.as_mut_ptr(), bytes.len());
             }
             self.buffer_usage = bytes.len();
         }
@@ -217,28 +215,12 @@ impl Hasher for XxHash {
     }
 }
 
-#[derive(Clone)]
-/// Constructs a randomized seed and reuses it for multiple hasher instances. See the usage warning on `XxHash32`.
-pub struct RandomXxHashBuilder(u32);
-
-impl RandomXxHashBuilder {
-    fn new() -> RandomXxHashBuilder {
-        RandomXxHashBuilder(rand::thread_rng().gen())
-    }
-}
-
-impl Default for RandomXxHashBuilder {
-    fn default() -> RandomXxHashBuilder { RandomXxHashBuilder::new() }
-}
-
-impl BuildHasher for RandomXxHashBuilder {
-    type Hasher = XxHash;
-
-    fn build_hasher(&self) -> XxHash { XxHash::with_seed(self.0) }
-}
+#[cfg(feature = "std")]
+pub use ::std_support::thirty_two::RandomXxHashBuilder;
 
 #[cfg(test)]
 mod test {
+    use std::prelude::v1::*;
     use std::hash::{Hasher, BuildHasherDefault};
     use std::collections::HashMap;
     use super::{XxHash, RandomXxHashBuilder};

